source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,258 @@ extern "C"
+            if ( i != col->column() )
+              {
+                *nextNborGID++ = col->column();
+                *nextNborProc++ = 0;    //All the vertices on processor 0
+              }
+        }
+    }
+
+
+    void partition_zoltan (const SparsityPattern     &sparsity_pattern,
+                           const unsigned int         n_partitions,
+                           std::vector<unsigned int> &partition_indices)
+    {
+      //MPI environment must have been initialized by this point.
+      Zoltan *zz = new Zoltan(MPI::COMM_SELF);
+
+      //Check if zoltan object is initialized
+      Assert( zz != nullptr , ExcEmptyObject() );
+
+      //General parameters
+      zz->Set_Param( "LB_METHOD", "GRAPH" );  //graph based partition method (LB-load balancing)
+      zz->Set_Param( "DEBUG_LEVEL", "0" );   //set level of debug info
+      zz->Set_Param( "NUM_LOCAL_PARTS", std::to_string(n_partitions) ); //set number of partitions
+
+      //Need a non-const object equal to sparsity_pattern
+      SparsityPattern graph;
+      graph.copy_from(sparsity_pattern);
+
+      //Set query functions
+      zz->Set_Num_Obj_Fn(get_number_of_objects, &graph);
+      zz->Set_Obj_List_Fn(get_object_list, &graph);
+      zz->Set_Num_Edges_Multi_Fn(get_num_edges_list, &graph);
+      zz->Set_Edge_List_Multi_Fn(get_edge_list, &graph);
+
+      //Variables needed by partition function
+      int changes=0;
+      int num_gid_entries=1;
+      int num_lid_entries=1;
+      int num_import=0;
+      ZOLTAN_ID_PTR import_global_ids=nullptr;
+      ZOLTAN_ID_PTR import_local_ids=nullptr;
+      int *import_procs=nullptr;
+      int *import_to_part=nullptr;
+      int num_export=0;
+      ZOLTAN_ID_PTR export_global_ids=nullptr;
+      ZOLTAN_ID_PTR export_local_ids=nullptr;
+      int *export_procs=nullptr;
+      int *export_to_part=nullptr;
+
+      //call partitioner
+      int rc = zz->LB_Partition (changes,
+                                 num_gid_entries,
+                                 num_lid_entries,
+                                 num_import,
+                                 import_global_ids,
+                                 import_local_ids,
+                                 import_procs,
+                                 import_to_part,
+                                 num_export,
+                                 export_global_ids,
+                                 export_local_ids,
+                                 export_procs,
+                                 export_to_part);
+
+      //check for error code in partitioner
+      Assert( rc == ZOLTAN_OK , ExcEmptyObject() );
+
+      Assert (partition_indices.size() != 0, ExcZero());
+      std::fill(partition_indices.begin(), partition_indices.end(), 0);
+
+      //copy from export_to_part to partition_indices
+      for (int i=0; i < num_export; i++)
+        {
+          partition_indices[export_local_ids[i]] = export_to_part[i];
```

Comment:
Is there a guarantee that `export_local_ids[i]` loops over all indices? If so, then we wouldn't need to fill the array by zeros immediately above. If there is no such guarantee, then filling the array with invalid indices might be the better choice.

Reply:
It is not intended to go through all the indices, rather only the ones that changed the partitions. This is why i set the default partition id to be zero.

Reply:
Hm, I don't think I understand. `partition_indices` is only an output variable -- as far as I can see, it's previous contents (if any) are not actually given to Zoltan, right? In that case, I would expect that Zoltan provides as its output a partition index for each node of the graph, and so `export_local_ids[i]` should loop through all nodes. If that were not the case, what partition is assigned to graph nodes that are not listed in `export_local_ids`? Right now, it is zero, but why is that the correct answer?

Reply:
Initially all of them belong to part 0. After partitioning, export_local_ids only has ids whose part ids(within each process) have changed. I will try if I can do it directly as you suggested!

Reply:
> After partitioning, export_local_ids only has ids whose part ids(within each process) have changed.

i think it's OK if you simply add this as a note above this line so that anyone reading the code would understand.

Reply:
p.s. from this code, it does not look like you can use `partition_indices` directly in Zoltan as the input is `ZOLTAN_ID_PTR export_local_ids=nullptr`. So it looks like Zoltan internally allocates memory and hopefully also cleans it up in destructor...

Reply:
That is true. Everything is allocated based on the query functions passed to zoltan

Reply:
Yes, just putting the reasoning into a comment is good enough -- simply state that Zoltan only provides information about nodes whose partition is != 0.